File: scripts/mandala_parity/parity_runner.mjs
// Parity-harness runner for Mandala.
//
// Reads JSON from stdin: {seed: int, actionChoices: [int, ...]}.
// Uses a mulberry32 PRNG seeded with `seed` to replace Math.random so
// the JS engine's shuffles match the Python side's mulberry32 bit-for-bit.
// Plays a game, taking the action at index actionChoices[t] from the
// legal-action list at turn t. Writes a JSON stream of per-step states
// (and final winner) to stdout.
//
// Invoke:  node parity_runner.js  (reads stdin, writes stdout).

import { readFileSync } from 'node:fs';
import {
  createGame,
  getValidActions,
  performAction,
  getWinner,
} from '../../../mandala-web/game.js';

// --- Patch Math.random to a seeded mulberry32 that matches the Python impl.
function installDeterministicRandom(seed) {
  let state = seed >>> 0;
  Math.random = function () {
    state = (state + 0x6D2B79F5) >>> 0;
    let t = state;
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= (t + Math.imul(t ^ (t >>> 7), t | 61)) >>> 0;
    return (((t ^ (t >>> 14)) >>> 0)) / 4294967296;
  };
}

function actionObjectsFromValid(valid) {
  const list = [];
  for (const a of valid.buildMountain) list.push({ type: 'build_mountain', ...a });
  for (const a of valid.growField) list.push({ type: 'grow_field', ...a });
  for (const a of valid.discardRedraw) list.push({ type: 'discard_redraw', ...a });
  for (const color of valid.claimColor) list.push({ type: 'claim_color', color });
  return list;
}

const input = JSON.parse(readFileSync(0, 'utf-8'));
const seed = input.seed >>> 0;
const choices = input.actionChoices;

installDeterministicRandom(seed);

let state = createGame('p0', 'p1');

const steps = [];
steps.push({ turn: 0, state });

for (let i = 0; i < choices.length; i++) {
  if (state.phase === 'ended') break;
  const valid = getValidActions(state);
  const flat = actionObjectsFromValid(valid);
  if (flat.length === 0) break;
  const idx = choices[i] % flat.length;
  const action = flat[idx];
  const res = performAction(state, action);
  if (!res.success) {
    steps.push({ turn: i + 1, error: res.error, state });
    break;
  }
  state = res.newState;
  steps.push({ turn: i + 1, action, state });
}

const winner = getWinner(state);
process.stdout.write(JSON.stringify({ steps, winner }));

File: scripts/mandala_parity/parity_runner.py
"""Python-side parity runner for Mandala. Mirrors parity_runner.js exactly.

Input/output format identical to the JS side so test harnesses can diff
them step-by-step. Both sides use the same mulberry32 seeded PRNG, so
shuffles and reshuffles produce identical deck orders.
"""

from __future__ import annotations

import json
import sys

from mlfactory.games.mandala._rng import Mulberry32
from mlfactory.games.mandala.rules import (
    create_game,
    get_valid_actions,
    get_winner,
    perform_action,
)


def action_objects_from_valid(valid: dict) -> list[dict]:
    """Flatten the getValidActions dict into a list of full action dicts.
    Order MUST match parity_runner.js exactly."""
    out: list[dict] = []
    for a in valid["buildMountain"]:
        out.append({"type": "build_mountain", **a})
    for a in valid["growField"]:
        out.append({"type": "grow_field", **a})
    for a in valid["discardRedraw"]:
        out.append({"type": "discard_redraw", **a})
    for color in valid["claimColor"]:
        out.append({"type": "claim_color", "color": color})
    return out


def main() -> None:
    payload = json.load(sys.stdin)
    seed = int(payload["seed"]) & 0xFFFFFFFF
    choices = list(payload["actionChoices"])

    rng = Mulberry32(seed)

    state = create_game("p0", "p1", rng=rng)

    steps = [{"turn": 0, "state": state}]

    for i, choice in enumerate(choices):
        if state["phase"] == "ended":
            break
        valid = get_valid_actions(state)
        flat = action_objects_from_valid(valid)
        if not flat:
            break
        idx = choice % len(flat)
        action = flat[idx]
        res = perform_action(state, action, rng=rng)
        if not res["success"]:
            steps.append({"turn": i + 1, "error": res["error"], "state": state})
            break
        state = res["newState"]
        steps.append({"turn": i + 1, "action": action, "state": state})

    winner = get_winner(state)
    sys.stdout.write(json.dumps({"steps": steps, "winner": winner}))


if __name__ == "__main__":
    main()
